python/graphstorm/dataloading/sampler.py

```diff
@@ -389,3 +392,298 @@ def sample_blocks(self, g, seed_nodes, exclude_eids=None):
+                self._index = -1
+            else:
+                # non-infinite sampler if it's for evaluation.
+                # set _index to -1 for next evaluation.
+                self._index = -1
+                return None
+
+        return self.file_indices[self._index]
+
+class RandomShuffleFileSampler():
+    r"""Random File Sampler. Has files reshuffled for sampling.
+
+    Parameters:
+    ----------
+    file_indices : list of int
+        File indices for a local trainer
+    """
+    def __init__(self, file_indices):
+        self.file_indices = file_indices
+        self.num_local_files = len(self.file_indices)
+        self._indices = list(range(self.num_local_files))
+
+    def __iter__(self):
+        self._index = -1
+        # shuffle the file indices
+        random.shuffle(self._indices)
+        return self
+
+    def __next__(self):
+        """ Get index for next file"""
+        self._index += 1
+
+        if self._index >= self.num_local_files:
+            # make it infinite sampler
+            random.shuffle(self._indices)
+            self._index = -1
+
+        return self.file_indices[self._indices[self._index]]
+
+class DistributedFileSampler(FileSamplerInterface):
+    r"""Distributed File Sampler. Samples file from the data path.
+    Each rank only has access to a partition of all the data shards.
+
+    Parameters:
+    ---------
+    dataset_path : str
+        path to the data files.
+        Mutually exclusive with :attr:`files`.
+    shuffle : bool
+        Set to ``True`` to have the files reshuffled
+        at every epoch. Shuffling is performed on the files of the local partition.
+    infinite : bool
+        Set to ``True`` to have the files sampled infinitely.
+    local_rank : int
+        Local rank ID.
+    world_size : int
+        Number of all trainers.
+    is_train : bool
+        Set to ``True`` if it's training set.
+    """
+    def __init__(
+        self, 
+        dataset_path=None, 
+        shuffle=False, 
+        infinite=False, 
+        local_rank=-1, 
+        world_size=1, 
+        is_train=True,
+    ):
+        super().__init__(dataset_path)
+
+        # Initialize distributed ranks
+        if dist.is_initialized():
+            self.world_size = 1
+            self.local_rank = 0
+        else:
+            self.local_rank = local_rank
+            self.world_size = world_size
+
+        # distribute file index
+        self._file_index_distribute()
+
+        if not is_train:
+            self.part_len = self.num_files
+        file_indices = list(range(self.part_len))
+        if shuffle and is_train:
+            sampler = RandomShuffleFileSampler(file_indices=file_indices)
+        else:
+            sampler = SequentialFileSampler(file_indices=file_indices, is_train=is_train)
+        self.sampler = sampler
+        self.sampler_iter = iter(sampler)
+
+        self.infinite = infinite
+        self.shuffle = shuffle
+
+    def _file_index_distribute(self):
```

Comment:
this is a lot of code and we don't have unit tests at all.

Reply:
added unit test.